ir/ir.rs

```diff
@@ -33,68 +33,66 @@ impl ItemsBuilder {
     /// Construct a new builder, with the given size.
     pub fn new(size: u32) -> ItemsBuilder {
         ItemsBuilder {
-            id_counter: 0,
             size,
             parsed: Default::default(),
             items: Default::default(),
+            data:  Default::default(),
             edges: Default::default(),
             roots: Default::default(),
         }
     }
 
-    fn next_id(&mut self) -> Id {
-        let id = Id(self.id_counter);
-        self.id_counter += 1;
-        id
-    }
-
     /// Add the given item to to the graph and return the `Id` that it was
     /// assigned.
-    pub fn add_item<T>(&mut self, key: &T, mut item: Item) -> Id {
-        let id = self.next_id();
-
-        item.id = id;
+    pub fn add_item(&mut self, item: Item) -> Id {
+        let id = item.id;
         self.items.insert(id, item);
 
-        let old_value = self.parsed.insert(key as *const T as *const (), id);
-        assert!(
-            old_value.is_none(),
-            "should not parse the same key into multiple items"
-        );
-
+        let old_value = self.parsed.insert(id);
+        assert!(old_value, "should not parse the same key into multiple items");
+        
         id
     }
 
     /// Add the given item to the graph as a root and return the `Id` that it
     /// was assigned.
-    pub fn add_root<T>(&mut self, key: &T, item: Item) -> Id {
-        let id = self.add_item(key, item);
+    pub fn add_root(&mut self, item: Item) -> Id {
+        let id = self.add_item(item);
         self.roots.insert(id);
         id
     }
 
     /// Add an edge between the given keys that have already been parsed into
     /// items.
-    pub fn add_edge<K, J>(&mut self, from: &K, to: &J) {
-        let from_id = self.id_for_key(from);
-        let to_id = self.id_for_key(to);
+    pub fn add_edge(&mut self, from: Id, to: Id) {
```

Comment:
This method should `debug_assert!` that we have items for the `from` and `to` ids. The `id_for_key` method used to do that, but its removal means that we could accidentally create dangling edges and not realize it until way down the line. Better to catch the bug early.